metals/src/main/scala/scala/meta/internal/metals/codeactions/ExtractValueCodeAction.scala

```diff
@@ -0,0 +1,311 @@
+package scala.meta.internal.metals.codeactions
+
+import scala.annotation.tailrec
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+import scala.meta.Defn
+import scala.meta.Enumerator
+import scala.meta.Name
+import scala.meta.Template
+import scala.meta.Term
+import scala.meta.Tree
+import scala.meta.inputs.Position
+import scala.meta.internal.metals.Buffers
+import scala.meta.internal.metals.CodeAction
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.parsing.Trees
+import scala.meta.internal.trees.Origin.Parsed
+import scala.meta.pc.CancelToken
+import scala.meta.tokens.Token
+
+import org.eclipse.lsp4j.CodeActionParams
+import org.eclipse.{lsp4j => l}
+
+class ExtractValueCodeAction(
+    trees: Trees,
+    buffers: Buffers
+) extends CodeAction {
+  override def kind: String = l.CodeActionKind.RefactorExtract
+
+  override def contribute(params: CodeActionParams, token: CancelToken)(implicit
+      ec: ExecutionContext
+  ): Future[Seq[l.CodeAction]] = Future {
+    val path = params.getTextDocument().getUri().toAbsolutePath
+    val range = params.getRange()
+
+    val applyOpt = trees.findLastEnclosingAt[Term.Apply](
+      path,
+      range.getStart(),
+      appl => appl.args.exists(_.pos.encloses(range))
+    )
+
+    val textEdits = for {
+      apply <- applyOpt
+      stats <- lastEnclosingStatsList(apply)
+      argument <- apply.args.find { arg => arg.pos.encloses(range) }
+      // avoid extracting lambdas (this needs actual type information)
+      if isNotLambda(argument)
```

Comment:
Ahhh, haha this was actually the first thing I tried when testing it.

Reply:
I need to do it via PresentationCompiler, doing some related work with inferring methods.